### PR TITLE
Extract storage into separate class; split Hashtable into Hashtable and Hashgraph.

### PR DESCRIPTION
**Pull request by ctb**
This PR separates the storage details of hashes from the actual hash function. It should support both more types of storage (4-bit and two byte CountMin sketches? maybe even exact storage?) and multiple hash functions.

It also separates out table and graph operations, making it easier to split class defn's between functionality that needs reversible hashing (graph operations) and functions that don't (table functions). Ultimately this can lead to new CPython objects (maybe Nodetable and Counttable?) that support k > 32 while the Nodegraph and Countgraph CPython objects would remain limited to k <= 32.

 I tried to use templates instead of composition but couldn't get the various friend & inheritance declarations to work.  Suggestions welcome.

No discernable performance impact was observed in some basic benchmarking.

Built on #1494. Might be a way to do #1490 (irreversible hashing for k > 32).
- [x] Is it mergeable?
- [x] `make test` Did it pass the tests?
- [x] `make clean diff-cover` If it introduces new functionality in
  `scripts/` is it tested?
- [x] `make format diff_pylint_report cppcheck doc pydocstyle` Is it well
  formatted?
- [x] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Is it documented in the `ChangeLog`?
  http://en.wikipedia.org/wiki/Changelog#Format
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?
- [x] Do the changes respect streaming IO? (Are they
  tested for streaming IO?)
- [x] Is the Copyright year up to date?

### CODE REVIEW

**Comment by ctb** [comment]
@betatim @luizirber I am having linker symbol resolution issues. Any suggestions?

**Comment by ctb** [comment]
Got it to compile!

Comments would be appreciated.  Is it worth following through on this and fixing up the load and save code?

**Comment by ctb** [comment]
All the tests pass!

@betatim @luizirber @standage ready for an initial review!

**Comment by ctb** [comment]
Ready for review! @betatim @standage

**Comment by betatim** [comment]
Mostly good 😄 

One pedantic comment: all the variables that used to be of type `Hashtable` but now are `Hashgraphs` are still called `hashtable` :-/ It'll start getting confusing when `Hashtable`s are used more why the variable called `hashtable` actually contains a graph...

I only spotted one used of `Hashtable` directly, everyone else goes through `Hashgraph`. Makes me wonder if the other `stop_bf` related things should continue to be `Hashtable`s as well. This is where I'll resume the reviewing.

File: lib/storage.hh [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+ */
+
+class BitStorage : public Storage
+{
+friend class Hashbits;
+protected:
+    std::vector<uint64_t> _tablesizes;
+    size_t _n_tables;
+    uint64_t _occupied_bins;
+    uint64_t _n_unique_kmers;
```
Comment (standage) [inline lib/storage.hh]: Why duplicate these class members? Isn't the idea of inheritance that the base class stores the common members and functions?

File: lib/storage.hh [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+ * tracking statistics, as well as save/load, and not much else.
+ *
+ */
+
+class BitStorage : public Storage
+{
+friend class Hashbits;
+protected:
+    std::vector<uint64_t> _tablesizes;
+    size_t _n_tables;
```
Comment (standage) [inline lib/storage.hh]: Isn't this redundant with `_tablesizes.size()`? Not a big deal, but why risk the two variables getting out of sync?

**Comment by ctb** [comment]
@standage, I'll fix your points on the next PR, which is forthcoming.

@betatim, the inheritance hierarchy will expand and I will do a more thorough review then.

Thanks all :).

**Comment by ctb** [comment]
On Thu, Nov 03, 2016 at 11:22:01AM -0700, Tim Head wrote:
> One pedantic comment: all the variables that used to be of type `Hashtable` but now are `Hashgraphs` are still called `hashtable` :-/ It'll start getting confusing when `Hashtable`s are used more why the variable called `hashtable` actually contains a graph...

this should all have been fixed in #1504.

**Pull request merged**